domain/state.go

```diff
@@ -99,6 +100,84 @@ func (st *StateBase) Prepare(query string, typeSamples ...any) (*sqlair.Statemen
 	return stmt, nil
 }
 
+// RunTx executes the closure function within the scope of a transaction.
+// The closure is passed a TxContext that can be passed on to state
+// functions, so that they can perform work within that same transaction.
+// The closure will be retried according to the transaction retry semantics,
+// if the transaction fails due to transient errors.
+// The closure should only be used to perform state changes and must not be used
+// to execute queries outside of the state scope.
+func (st *StateBase) RunTx(ctx context.Context, fn func(TxContext) error) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Annotate(err, "getting database")
+	}
+
+	return db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// The txContext is created with the transaction and passed to the
+		// closure function. This ensures that the transaction is always
+		// available to the closure. Once the transaction is complete, the
+		// transaction is removed from the txContext. This is to prevent
+		// the transaction from being used outside of the transaction scope.
+		// This will prevent any references to the sqlair.TX from being held
+		// outside of the transaction scope.
+
+		// TODO (stickupkid): The txContext can be pooled on the StateBase to
+		// reduce the number of allocations. Attempting to push the tx into
+		// the context would prevent that as a viable option.
+		txCtx := &txContext{
+			Context: ctx,
+			t:       tx,
+		}
+		defer txCtx.removeTx()
+
+		return fn(txCtx)
+	})
+}
+
+// TxStateBase is an interface that provides a method for executing a closure
+// within the scope of a transaction.
+type TxStateBase interface {
+	// RunTx executes the closure function within the scope of a transaction.
+	// The closure is passed a TxContext that can be passed on to state
+	// functions, so that they can perform work within that same transaction.
+	// The closure will be retried according to the transaction retry semantics,
+	// if the transaction fails due to transient errors.
+	// The closure should only be used to perform state changes and must not be
+	// used to execute queries outside of the state scope.
+	RunTx(ctx context.Context, fn func(TxContext) error) error
+}
+
+// Run executes the closure function using the provided TxContext as
+// the transaction context.
+// It is expected that the closure will perform state changes within the
+// transaction scope.
+// Any errors returned from the closure are coerced into a standard error
+// to prevent sqlair errors from being returned to the Service layer.
+func Run(ctx TxContext, fn func(context.Context, *sqlair.TX) error) error {
+	txCtx, ok := ctx.(*txContext)
+	if !ok {
+		// If you're seeing this error, it means that the TxContext was not
+		// created by RunTx. This is a programming error. Did you attempt to
+		// wrap the context in a custom context and pass it to Run?
+		return fmt.Errorf("programmatic error: TxContext is not a *txContext: %T", ctx)
```

Comment:
One thing I had considered, if `ctx` is not a `txContext`, should we just start a transaction for this call?

My initial thought here is that it is strange there are some state methods that are for direct calling (e.g. existing methods that call `db.Txn`) and some that need to be wrapped in a closure (e.g. compose-able state operations that don't start their own txn).

I'm not saying that all state methods can have a call site in `RunTx`, I'm just saying that it is safe for all state operations to start their own txns.

But then it leads me to this question:
- Does it make sense then to have any state operation that starts its own transaction?
If there are some state operations that start their own txn and some that don't, it will lead to these paths:
1. Duplication of code; -or-
2. Lack of compos-ability; -or-
3. No one using the current form of state methods.

The issue with mixing the two styles of operations in the same interfaces is that one cannot and must not call a state method that starts its own txn. Isn't that possible to have a deadlock then?

I guess what I'm suggesting is that we go all in on this after it lands.

Reply:
We discussed this in yesterday's meeting. I believe it was deemed undesirable to start a transaction or not based on the particular context passed.

I don't mind having state methods that still call `db.Txn`. We disagreed somewhat on this. Simon argued against it; Ian and I were OK with it. If you have a service function only returning data, have it as a one-liner, and use `db.Txn` in the state call.

We already have some state methods that receive a transaction in order that they can be used in multiple places; those will still be suitable for calling with `RunTx`.